### PR TITLE
feat(table): `pivot_longer` API

### PR DESCRIPTION
**Pull request by cpcloud**
This is a PR to discuss a [`pivot_longer`](https://tidyr.tidyverse.org/reference/pivot_longer.html)-style API for table expressions. Ideally this can go into 5.0, but the main purpose is to generate discussion around working code. Depends on #5662.

### CODE REVIEW

**Comment by thegreymatter** [comment]
I love the portable implementation approach, but it raises a questions - 
since many backends support native impl (which might, or might not be more optimised, but probably more readable), is there a way to allow backend specific impl, while keeping the portable impl as a fallback?

I think it is possible if implementing the BaseBackend version with the base expression.

**Comment by cpcloud** [comment]
> I love the portable implementation approach, but it raises a questions - since many backends support native impl (which might, or might not be more optimised, but probably more readable), is there a way to allow backend specific impl, while keeping the portable impl as a fallback?
> 

Totally!

The current way I'm implementing it also very slow for the DuckDB backend in particular. Implementing it by [using `UNNEST`](https://github.com/duckdb/duckdb/pull/6387) makes the generic implementation much much faster.

The main problem with the generic approach using unnest is that many fewer backends support arrays and unnest than support union alls.

> I think it is possible if implementing the BaseBackend version with the base expression.

Yeah, I think for now that's probably a reasonable approach until we have a way to dispatch per-backend at the expression level and not just the operation level.

File: ibis/expr/types/relations.py [hunk 1]
```diff
@@ -2319,6 +2321,366 @@ def to_pandas(self, **kwargs) -> pd.DataFrame:
...
+        │ Afghanistan │ AF     │ AFG    │  1980 │ sp        │      1 │    4 │     ∅ │
+        │ Afghanistan │ AF     │ AFG    │  1980 │ sp        │      1 │    5 │     ∅ │
+        │ Afghanistan │ AF     │ AFG    │  1980 │ sp        │      1 │    6 │     ∅ │
+        │ Afghanistan │ AF     │ AFG    │  1980 │ sp        │      2 │    0 │     ∅ │
+        │ Afghanistan │ AF     │ AFG    │  1980 │ sp        │      2 │    1 │     ∅ │
+        │ Afghanistan │ AF     │ AFG    │  1980 │ sp        │      2 │    2 │     ∅ │
+        │ …           │ …      │ …      │     … │ …         │      … │    … │     … │
+        └─────────────┴────────┴────────┴───────┴───────────┴────────┴──────┴───────┘
+
+        `names_transform` can to be:
```
Comment (jcrist) [inline ibis/expr/types/relations.py:2557]: typo

```suggestion
        `names_transform` can be:
```

Comment (cpcloud) [reply ibis/expr/types/relations.py:2557]: Done.

File: ibis/expr/types/relations.py [hunk 2]
```diff
@@ -2319,6 +2321,366 @@ def to_pandas(self, **kwargs) -> pd.DataFrame:
...
+
+            1. A mapping of one or more names in `names_to` to callable
+            2. A callable that will be applied to every names
+            3. An iterable of callables that will be applied to the values from
+               `names_to` in order. Any `None` values are interpreted as "no
+               transformation".
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/expr/types/relations.py:2563]: Let's keep it as single callable or mapping of name -> callable for now

File: ibis/expr/types/relations.py [hunk 3]
```diff
@@ -2319,6 +2321,366 @@ def to_pandas(self, **kwargs) -> pd.DataFrame:
...
+        │ Afghanistan │ AF     │ AFG    │  1980 │ sp        │      2 │    0 │     ∅ │
+        │ Afghanistan │ AF     │ AFG    │  1980 │ sp        │      2 │    1 │     ∅ │
+        │ Afghanistan │ AF     │ AFG    │  1980 │ sp        │      2 │    2 │     ∅ │
+        │ …           │ …      │ …      │     … │ …         │      … │    … │     … │
+        └─────────────┴────────┴────────┴───────┴───────────┴────────┴──────┴───────┘
+
+        `names_transform` can to be:
+
+            1. A mapping of one or more names in `names_to` to callable
+            2. A callable that will be applied to every names
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/expr/types/relations.py:2560]: Move the description up and deleted the example.

File: ibis/backends/tests/test_generic.py [hunk 4]
```diff
@@ -937,3 +938,37 @@ def query(t, group_cols):
...
+    raises=com.OperationNotDefinedError,
+)
+@pytest.mark.notyet(
+    ["datafusion", "impala", "mssql", "mysql", "sqlite"],
+    reason="backend doesn't support arrays and we don't implement pivot_longer with unions yet",
+    raises=com.OperationNotDefinedError,
+)
+def test_pivot_longer(con):
+    diamonds = con.tables.diamonds
+    res = diamonds.pivot_longer(s.c("x", "y", "z"), names_to="pos", values_to="xyz")
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/tests/test_generic.py:961]: Will do!

File: ibis/expr/types/relations.py [hunk 5]
```diff
@@ -2481,6 +2484,314 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            transform a column name to a value.
+        values_to
+            Name of the pivoted value column.
+        values_transform
+            Apply a function to the value column. This can be a lambda or
+            deferred expression.
+
+        Returns
+        -------
+            Pivoted table
```
Comment (jcrist) [inline ibis/expr/types/relations.py:2522]: Should this be indented like this? I'm used to the numpydoc style, not sure what our mkdocs api docs formats nicely.

File: ibis/expr/types/relations.py [hunk 6]
```diff
@@ -2481,6 +2484,314 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+        ┃ country ┃ indicator   ┃ year   ┃ value   ┃
+        ┡━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━┩
+        │ string  │ string      │ string │ float64 │
+        ├─────────┼─────────────┼────────┼─────────┤
+        │ ABW     │ SP.URB.TOTL │ 2000   │ 42444.0 │
+        │ ABW     │ SP.URB.TOTL │ 2001   │ 43048.0 │
+        │ ABW     │ SP.URB.TOTL │ 2002   │ 43670.0 │
+        │ ABW     │ SP.URB.TOTL │ 2003   │ 44246.0 │
+        │ ABW     │ SP.URB.TOTL │ 2004   │ 44669.0 │
+        └─────────┴─────────────┴────────┴─────────┘
```
Comment (jcrist) [inline ibis/expr/types/relations.py:2593]: Would you mind adding a sentence or two describing what's done in the two examples above. This is both to provide more context into what `pivot_longer` is doing, and to also add a break between the two examples above, which kinda blend together without a sentence to break them apart.

**Review by jcrist — APPROVED** [review]
One question, and one small nit, otherwise this looks good to me!

File: ibis/expr/types/relations.py (hunk 5 above)
Comment (cpcloud) [reply ibis/expr/types/relations.py:2522]: Ha! This is a rebase error on my part. Thanks for catching.

File: ibis/expr/types/relations.py (hunk 6 above)
Comment (cpcloud) [reply ibis/expr/types/relations.py:2602]: Done, filled out the rest of the examples with more prose.

File: ibis/expr/types/relations.py (hunk 5 above)
Comment (cpcloud) [reply ibis/expr/types/relations.py:2523]: Fixed!

**Pull request merged**